### PR TITLE
feat: instrument controlplane with metrics

### PR DESCRIPTION
**Pull request by AndreasZeissner**
## Motivation and Context

This introduces metrics for the controlplane by adding a local fastify plugin.
The plugin uses prom-client to collect the default metrics provided by `collectDefaultMetrics`.

The following environment variables are used to enable metrics in the controlplane:

`PROMETHEUS_ENABLED`: registers the metrics plugin (defaults to true)
`PROMETHEUS_HTTP_PATH`: on which path to expose the metrics (defaults to '/metrics'
`PROMETHEUS_PORT`: the port to listen on for the metrics listener (defaults to 8088)
`PROMETHEUS_HOST`: the host to listen to (defaults to localhost)

To enable the metrics endpoint deployed through helm ensure the following values of the controlplane helm chart are set properly:

```
  # Use this section to disable/enable and configure prometheus metrics.
  prometheus:
    # -- Enables prometheus metrics support. Default is false.
    enabled: false
    # -- The host to bind to defautls to 127.0.0.1 to avoid opening the metrics endpoint by default.
    host: 127.0.0.1
    # -- The port where metrics are exposed. Default is port 8088.
    port: 8088
    # -- The HTTP path where metrics are exposed. Default is "/metrics".
    path: "/metrics"
```

By default, metrics are not exposed from the controlplane.
This is done to avoid opening sensitive information to the public by accident.

The metrics include e.g.:

```
process_cpu_user_seconds_total
process_cpu_system_seconds_total
process_cpu_seconds_total
process_start_time_seconds
process_resident_memory_bytes
process_virtual_memory_bytes
process_heap_bytes
process_open_fds
nodejs_heap_space_size_available_bytes{space="read_only"}
nodejs_heap_space_size_available_bytes{space="new"}
nodejs_heap_space_size_available_bytes{space="old"}
nodejs_heap_space_size_available_bytes{space="code"}
nodejs_heap_space_size_available_bytes{space="shared"}
nodejs_heap_space_size_available_bytes{space="new_large_object"}
nodejs_heap_space_size_available_bytes{space="large_object"}
nodejs_heap_space_size_available_bytes{space="code_large_object"}
nodejs_heap_space_size_available_bytes{space="shared_large_object"}
```

See for further information: [Prom Client Metrics](https://github.com/siimon/prom-client/tree/master/lib/metrics)

## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: controlplane/src/core/plugins/metrics.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  const collectDefaultMetrics = client.collectDefaultMetrics;
+  collectDefaultMetrics({});
+
+  await listener.route({
+    method: 'GET',
+    url: path,
+    handler,
+  });
+
+  await listener.listen({
```
Comment (StarpTech) [inline controlplane/src/core/plugins/metrics.ts:35]: You have to close the server on the fastify close hook.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: controlplane/src/core/plugins/metrics.ts (hunk 1 above)
Comment (AndreasZeissner) [reply controlplane/src/core/plugins/metrics.ts:41]: Can you have a look if this is now as extended.

The onClose hook of the controlplane process will ensure that the server is closed as well.

File: controlplane/src/core/plugins/metrics.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+export interface MetricsPluginOptions {
+  enabled?: boolean;
+  logger?: Logger;
+  path: string;
+  port: number;
+  host: string;
+}
+
+const handler = async function (request: FastifyRequest, reply: FastifyReply) {
+  const metrics = await client.register.metrics();
```
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply controlplane/src/core/plugins/metrics.ts]: It's now using a dedicated registry in the metrics endpoint.

File: controlplane/src/core/plugins/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  reply.send(metrics);
+};
+
+export default fp<MetricsPluginOptions>(async function (fastify, opts) {
+  const { host, port, path, logger } = opts;
+  const listener = Fastify({
+    logger,
```
[the remark this answers is not in the document]
Comment (AndreasZeissner) [reply controlplane/src/core/plugins/metrics.ts:25]: Let me know if the module string  should be something else:
```
[16:40:19 UTC] INFO: Server listening at http://127.0.0.1:8088 {"module":"metrics"}
[16:40:19 UTC] INFO: Server listening at http://127.0.0.1:3001
```

[the remark this answers is not in the document]
Comment (StarpTech) [reply controlplane/src/core/plugins/metrics.ts]: Looks good

File: controlplane/src/core/plugins/metrics.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (StarpTech) [reply controlplane/src/core/plugins/metrics.ts]: But it was still global. I modified it in a way that it always create a separate one per test.

File: controlplane/src/core/plugins/metrics.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+export interface MetricsPluginOptions {
+  enabled?: boolean;
+  registry?: Registry;
+  path?: string;
+}
+
+export default fp<MetricsPluginOptions>(function (fastify, { path = '/metrics', registry = new Registry() }, next) {
+  const metricsLogger = fastify.log.child({
+    component: '@wundergraph/controlplane',
```
Comment (StarpTech) [inline controlplane/src/core/plugins/metrics.ts:21]: I like the idea but we haven't done this somewhere else on the controlplane. I suggest to align it everywhere or leave it for now. Also logs are already scoped to a particular service. We wouldn't win any context by flagging logs that they come from the service.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: helm/cosmo/charts/controlplane/values.yaml [hunk 5]
```diff
@@ -191,5 +191,24 @@ configuration:
...
+    # -- Enables prometheus metrics support. Default is false.
+    enabled: false
+    # -- The host to bind to defautls to 127.0.0.1 to avoid opening the metrics endpoint by default.
+    host: 127.0.0.1
+    # -- The port where metrics are exposed. Default is port 8088.
+    port: 8088
+    # -- The HTTP path where metrics are exposed. Default is "/metrics".
+    path: "/metrics"
+    # Use this section to disable/enable pod monitoring over [GCP managed collection](https://cloud.google.com/stackdriver/docs/managed-prometheus/setup-managed#config-mgd-collection).
+    gcp:
```
Comment (StarpTech) [inline helm/cosmo/charts/controlplane/values.yaml:205]: Let's call this: `gcp_managed_prometheus`

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**